### PR TITLE
Added WHY operations and new parameters for details on entities

### PR DESCRIPTION
**Pull request by barrycaceres**
Added support for why functions at the following endpoints:

- GET /data-sources/{dataSourceCode}/records/{recordId}/entity/why

- GET /entities/{entityId}/why

- GET /why/records?dataSource1={src1}&recordId1={id1}&dataSource2={src2}&recordId2={id2}

Also added support for the "withFeatureStatistics" and "withDerivedFeatures" parameters across the following endpoints:

- GET /data-sources/{dataSourceCode}/records/{recordId}/entity

- GET /data-sources/{dataSourceCode}/records/{recordId}/entity/why

- GET /entities/{entityId

- GET /entities/{entityId}/why

- GET /why/records?dataSource1={src1}&recordId1={id1}&dataSource2={src2}&recordId2={id2}

- GET /entity-paths

- GET /entity-networks

Finally, "featureDetails" property was added to entity results to support obtaining the feature ID as well as the feature statistics (if requested).

### CODE REVIEW

File: src/main/java/com/senzing/api/BuildInfo.java [hunk 1]
```diff
@@ -7,7 +7,7 @@
 public class BuildInfo {
   public static final String MAVEN_VERSION;
 
-  public static final String REST_API_VERSION = "1.8.0";
+  public static final String REST_API_VERSION = "1.8.1";
```
Comment (docktermj) [inline src/main/java/com/senzing/api/BuildInfo.java:10]: Should this be `1.8.2`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,16 +8,34 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [1.8.1] - 2020-03-30
```
Comment (docktermj) [inline CHANGELOG.md:9]: Looks like this should be 

```
## [1.8.2] - 2020-04-15
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,16 +8,34 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  - GET /data-sources/{dataSourceCode}/records/{recordId}/entity
+  - GET /data-sources/{dataSourceCode}/records/{recordId}/entity/why
+  - GET /entities/{entityId
+  - GET /entities/{entityId}/why
+  - GET /why/records
+  - GET /entity-paths
+  - GET /entity-networks
+- Added the "featureDetails" property to entity results to support obtaining
+  the feature ID as well as the feature statistics (if requested).
+
```
Comment (docktermj) [inline CHANGELOG.md:28]: The 1.8.1 section should remain here.
```
## [1.8.1] - 2020-03-30
```

File: src/main/java/com/senzing/api/model/SzCandidateKey.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+   * for the Senzing native API format for a candidate key to create a new
+   * instance.
+   *
+   * @param jsonObject The {@link JsonObject} to parse.
+   *
+   * @param featureType The feature type for the {@link SzCandidateKey}.
+   *
+   * @return The {@link SzCandidateKey} that was created.
+   */
+  public static SzCandidateKey parseCandidateKey(JsonObject jsonObject,
```
Comment (timadair) [inline src/main/java/com/senzing/api/model/SzCandidateKey.java:103]: Between the class name and the return type there's no need to include the type in the name in the static methods.  SzCandidateKey.parse(...) conveys just as much information with less noise.

File: src/main/java/com/senzing/api/model/SzCandidateKey.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  {
+    if (list == null) {
+      list = new ArrayList<>(jsonArray.size());
+    }
+
+    for (JsonObject jsonObject : jsonArray.getValuesAs(JsonObject.class)) {
+      list.add(parseCandidateKey(jsonObject, featureType));
+    }
+
+    return list;
```
Comment (timadair) [inline src/main/java/com/senzing/api/model/SzCandidateKey.java:169]: Having a mutable list be both a parameter and a return type seems very awkward.  I'd rather see it be one or the other.  Either have it be passed in and fully commit to the side effects, or remove the list parameter and then addAll() to whatever list you're adding to.

File: src/main/java/com/senzing/api/model/SzCandidateKey.java (hunk 4 above)
Comment (docktermj) [reply src/main/java/com/senzing/api/model/SzCandidateKey.java:103]: I like your thinking, @timadair

File: src/main/java/com/senzing/api/model/SzEntityFeatureDetail.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+   * @return The actual feature value.
+   */
+  public String getFeatureValue() {
+    return featureValue;
+  }
+
+  /**
+   * Sets the actual feature value.
+   *
+   * @param featureValue The actual feature value.
```
Comment (timadair) [inline src/main/java/com/senzing/api/model/SzEntityFeatureDetail.java:71]: Is there a fake feature value?  What's this supposed to mean?

File: src/main/java/com/senzing/api/services/WhyServices.java [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+      String rawData = null;
+
+      int flags = getFlags(forceMinimal,
+                           featureMode,
+                           withFeatureStats,
+                           withDerivedFeatures,
+                           withRelationships);
+
+      enteringQueue(timers);
+      rawData = provider.executeInThread(() -> {
```
Comment (timadair) [inline src/main/java/com/senzing/api/services/WhyServices.java:65]: This lambda looks like it's being duplicated in several methods, and the only difference is the string describing the method name and the method to be called.  Is there a reason why you couldn't extract this?

**Pull request merged**